docs/RFCS/time_series_culling.md

```diff
@@ -0,0 +1,349 @@
+cluster performance over larger times scales, without requiring the storage of
+an excessive amount of metrics data.
+
+# Detailed design
+
+## Culling algorithm
+
+The culling algorithm is responsible for identifying high-resolution time series
+keys that are older than a system-set threshold. Once identified, the keys are
+passed into the rollup/delete process.
+
+There are two primary design requirements of the culling algorithm:
+
+1. From a single node, efficiently locating time series keys which need to be 
+culled.
+2. Across the cluster, efficiently distributing the task of culling with minimal
+coordination between nodes.
+
+#### Locating Time Series Keys
+
+Locating time series keys to be culled is not completely trivial due to the
+construction of time series keys, which is thus: 
+`[ts prefix][series name][timestamp][source]`
+
+> Example: "ts/cr.node.sql.inserts/1473739200/1" would contain time series data
+> for "cr.node.sql.inserts" on September 13th 2016 between 4am-5am UTC,
+> specifically for node 1.
+
+Because of this construction, which prioritizes name over timestamp, the most 
+recent time series data for series "A" would sort *before* the oldest time
+series data for series "B". This means that we cannot simply cull the beginning
+of the time series range.
+
+The simplest alternative would be to scan the time series range looking for
+invalid keys; however, this is considered to be a burdensome scan due to the
+number of keys that are not culled. For a per-node time series being recorded on
+a 10 node cluster with a 2 week retention period, we would expect to retain (10
+x 24 x 14) = *3360* keys that should not be culled. In a system that maintains
+dozens, possibly hundreds of time series, this is a lot of data for each node to
+scan on a regular basis.
+
+However, this scan can be effectively distributed across the cluster by creating
+a new *replica queue* which searches for time series keys. The new queue can
+quickly determine if each range contains time series keys (by inspecting 
+start/end keys); for ranges that do contain time series keys, specific keys
+can then be inspected at the engine level. This means that key inspections do
+not require network calls, and the number of keys that can be inspected at once
+is limited to the size of a range.
+
+Once the queue discovers a range that contains time series keys, the scanning
+process does not need to inspect every key on the range. The algorithm is as
+follows:
+
+1. Find the first time series key in the range (scan for [ts prefix]). 
+2. Deconstruct the key to retrieve its name.
+3. Run the rollup/delete operation on all keys in the range: 
+    `[ts prefix][series name][0] - [tsprefix][series name][now - threshold]`
+4. Find the next key on the range which contains data for a different time
+series by searching for key `PrefixEnd([ts prefix][series name])`.
+5. If a key was found in step 5, return to step 2 with that series name.
```

Comment:
Judging from the beginnings of this in #9767, is this what would happen in the queue's `process`?

- find all the time series names (not only the first one) from the engine
- Iterating over them, generating and sending out for each a batch which contains the Rollout and the DeleteRange?

I'm asking because the `Rollout` and the `DeleteRange` should be tied together as closely as possible, and that (at least potentially) forces `DistSender` to put them into a transaction (due to spanning ranges). This necessitates intents on the timeseries keys, which is a whole can of worms we should stay away from.

So instead of that, you are going to have to send two individual batches, presumably doing the rollup first, and then the deletion. I think that's ok (though see comments below), but wanted to point it out as that's a good one to waste time on during imlementation.